README
hbackup
=======

hbackup is a tool for efficiently backing up a UNIX filesystem to
another filesystem (for instance on a removable hard disk).  It works
on Linux and Mac OS X.

There are currently two command tools provided: a Python script called
hbackup, and a C++ program called nhbackup.


Installation
------------

   make
   make install         # as root

See 'man hbackup' for documentation.
   

How it works
------------

Backups are performed by writing an index of all the directories and
files on the source filesystem to a single file.  This records their
name, permissions, etc.  For very short files it may also include
their contents but for long files it just includes a hash of the
contents.

There is a separate directory structure which contains a mapping
between hashes and file contents.

Some useful consequences of this design are:

 1) Duplicate files are stored only once on the backup medium.  When
    the second copy is encountered, it is already there in the hash
    directory and so need not be written again.

    Moreover if you back up several hosts to the same medium then
    files common to both hosts, for instance many of the files under
    /usr, only appear once on the backup device.

 2) All backups to the same medium are incremental, for the same
    reason.  When you make a second backup, only the index and new
    files need b e written to the backup medium.

 3) The demands on the backup filesystem are relatively light, and
    FAT32 will suffice.  This makes it easier to share backup devices
    between different platforms.

It should be said that a less useful consequence is that when you
delete old indexes, the removal of files no longer referenced by any
index is a relatively slow and expensive process.


How to use it
-------------

My nightly backups are organized as follows:

   * I have a USB hard disk mounted on /usb.  It's formatted as FAT32.

   * Local backups use the script backup.sfere, run nightly from cron.

   * Remote backups from other hosts use the various rbackup.*
     scripts.  These are invoked by SSH from the machine with the
     backup device, to ensure that only backup is running at once.

See the man page, hbackup.1, for command-line options etc.


Technical Difficulties
----------------------

1. SFTP as implemented by OpenSSH (at the time of writing) does not
   support renames on FAT32 filesystems.  This is because it tries to
   provide a non-overwriting rename with link() and unlink() rather
   than an overwriting rename with rename() (which is what hbackup
   actually wants).

   The Python script hbackup has a --no-rename option which writes
   hash files in place but this is less than satisfactory.  A better
   alternative if backing up over SFTP is to apply openssh.patch and
   install the resulting SFTP server in /usr/local/bin, and use the
   --sftp-server option to nhbackup to access it.

   A further alternative is to use an SFTP server that supports the
   posix-rename extension.

2. Mac OS X is capable of returning filenames from readdir() than
   cannot then be accessed.  Currently hbackup will fail entirely if
   this happens.  nhbackup will skip such files.

   Neither of these are very good, but it's not clear what alternative
   there is.


Mailing Lists
-------------

http://www.chiark.greenend.org.uk/mailman/listinfo/sgo-software-discuss
 - discussion, bug reports, etc

http://www.chiark.greenend.org.uk/mailman/listinfo/sgo-software-announce
 - announcements of new versions


Author
------

hbackup is (c) 2006, 2007 Richard Kettlewell.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


CHANGES
Changes in version 0.2
======================

   * The order of index files is now normalised, making it easier to
     compare them.

   * New --hint-file option uses filesystem information to record
     hashes, removing the need to rehash large files.  This can
     significantly speed up some backups.

   * When backing up, hashes are checked for in the repository
     asynchronously.

   * When copying a file into the repo, the data actually written is
     hashed, to detect "underfoot" changes.  The check can be
     suppressed by --no-recheck-hash, for the case where you know
     there will be no change.

   * Various bug fixes.


local.cc
/*
 * This file is part of hbackup.
 * Copyright (C) 2006, 2007 Richard Kettlewell
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 */
#include "nhbackup.h"

// Local Filesystem -----------------------------------------------------------

LocalFilesystem local;

LocalFile::LocalFile(const string &path_, OpenMode mode) :
  fd(-1), path(path_) {
  mode_t m;
  
  switch(mode) {
  case ReadOnly: m = O_RDONLY; break;
  case Overwrite: m = O_WRONLY|O_CREAT|O_TRUNC; break;
  case NoOverwrite: m = O_WRONLY|O_CREAT|O_EXCL; break;
  default: fatal("invalid open mode %d", (int)mode);
  }
  if((fd = open(path_.c_str(), m, 0666)) < 0)
    throw FileError("opening", path, errno);
  fprintf(stderr, "opened a localfile %s\n", path.c_str());
}

LocalFile::~LocalFile() {
  if(fd != -1 && close(fd) < 0) throw FileError("closing", path, errno);
}
 
int LocalFile::readbytes(void *buf, int space) {
  int n;
  if((n = read(fd, buf, space)) < 0) throw FileError("reading", path, errno);
  return n;
}

void LocalFile::writebytes(const void *buf, int nbytes) {
  int n, written = 0;

  fprintf(stderr,  "write %d bytes to %s\n", nbytes, path.c_str());
  while(written < nbytes) {
    if((n = write(fd, (const char *)buf + written, nbytes)) < 0)
      throw FileError("writing", path, errno);
    written += n;
  }
}

bool LocalFile::readable() const {
  fd_set fds;
  assert(mode != writing);
  if(next != top) return true;          // buffered data, can definitely read
  FD_ZERO(&fds);
  FD_SET(fd, &fds);
  struct timeval tv = { 0, 0 };
  int n;
  do {
    n = select(fd + 1, &fds, 0, 0, &tv);
  } while(n < 0 && errno == EINTR);
  if(n < 0) fatal("select: %s", strerror(errno));
  return !!FD_ISSET(fd, &fds);
}

void LocalFilesystem::rename(const string &oldpath, const string &newpath) {
  if(::rename(oldpath.c_str(), newpath.c_str()) < 0)
    throw FileError("renaming", oldpath, errno);
}

void LocalFilesystem::link(const string &oldpath, const string &newpath) {
  if(::link(oldpath.c_str(), newpath.c_str()) < 0)
    throw FileError("linking", oldpath, errno);
}
  
void LocalFilesystem::remove(const string &path) {
  if(::remove(path.c_str()) < 0) 
    throw FileError("removing", path, errno);
}

File *LocalFilesystem::open(const string &path, OpenMode mode) {
  return new LocalFile(path, mode);
}

void LocalFilesystem::mkdir(const string &path, mode_t mode) {
  if(::mkdir(path.c_str(), mode) < 0)
    throw FileError("creating directory", path, errno);
}

void LocalFilesystem::mknod(const string &path, mode_t mode, dev_t dev) {
  if(::mknod(path.c_str(), mode, dev) < 0)
    throw FileError("creating device", path, errno);
}

void LocalFilesystem::lchown(const string &path, uid_t uid, gid_t gid) {
  if(::lchown(path.c_str(), uid, gid) < 0)
    throw FileError("changing ownership of", path, errno);
}

void LocalFilesystem::chmod(const string &path, mode_t mode) {
  if(::chmod(path.c_str(), mode) < 0)
    throw FileError("changing permissions of", path, errno);
}

void LocalFilesystem::symlink(const string &target, const string &path) {
  if(::symlink(target.c_str(), path.c_str()) < 0)
    throw FileError("creating symlink", path, errno);
}

void LocalFilesystem::utimes(const string &path, time_t atime, time_t mtime) {
  struct timeval times[2];
  
  times[0].tv_sec = atime;
  times[0].tv_usec = 0;
  times[1].tv_sec = mtime;
  times[1].tv_usec = 0;
  if(::utimes(path.c_str(), times) < 0)
     throw FileError("setting file times", path, errno);
}

int LocalFilesystem::exists(const string &path) {
  struct stat sb;

  if(stat(path.c_str(), &sb) < 0) {
    if(errno != ENOENT)
      throw FileError("stat", path, errno);
    return 0;
  }
  return 1;
}

void LocalFilesystem::contents(const string &path,
			       list<string> &c) {
  DIR *dp;
  struct dirent *de;

  c.clear();
  if(!(dp = opendir(path.c_str())))
    throw FileError("opening directory", path, errno);
  try {
    errno = 0;
    while((de = readdir(dp))) {
      string s = de->d_name;
      if(s != "." && s != "..")
	c.push_back(string(s));
      errno = 0;
    }
    if(errno) throw FileError("reading directory", path, errno);
  } catch(...) {
    closedir(dp);
    throw;
  }
  if(closedir(dp) < 0) throw FileError("closing directory", path, errno);
}

string LocalFilesystem::readlink(const string &path) {
  char buffer[MAXLINKSIZE];

  int n = ::readlink(path.c_str(), buffer, sizeof buffer);
  if(n < 0)
    throw FileError("reading link", path, errno);
  if((unsigned)n >= sizeof buffer)
    throw FileError("reading link", path, ENAMETOOLONG);
  buffer[n] = 0;
  return buffer;
}

bool LocalFilesystem::ismount(const string &path) {
  struct stat sb, psb;
  const string parent = path + "/..";

  if(stat(path.c_str(), &sb) < 0) throw FileError("stat", path, errno);
  if(!S_ISDIR(sb.st_mode)) return false;
  if(stat(parent.c_str(), &psb) < 0) throw FileError("stat", parent, errno);
  return sb.st_dev != psb.st_dev;
}

Filetype LocalFilesystem::type(const string &path) {
  struct stat sb;

  if(lstat(path.c_str(), &sb) < 0)
    throw FileError("lstat", path, errno);
  if(S_ISREG(sb.st_mode)) return RegularFile;
  else if(S_ISDIR(sb.st_mode)) return Directory;
  else if(S_ISLNK(sb.st_mode)) return SymbolicLink;
  else return UnknownFileType;
}

/*
Local Variables:
c-basic-offset:2
comment-column:40
fill-column:79
indent-tabs-mode:nil
End:
*/
